flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAligner.java

```diff
@@ -100,21 +105,20 @@ public boolean isBlocked(int channelIndex) {
 	}
 
 	@Override
-	public boolean processBarrier(CheckpointBarrier receivedBarrier, int channelIndex, long bufferedBytes) throws Exception {
+	public void processBarrier(CheckpointBarrier receivedBarrier, int channelIndex) throws Exception {
 		final long barrierId = receivedBarrier.getId();
 
 		// fast path for single channel cases
 		if (totalNumberOfInputChannels == 1) {
 			if (barrierId > currentCheckpointId) {
 				// new checkpoint
 				currentCheckpointId = barrierId;
-				notifyCheckpoint(receivedBarrier, bufferedBytes, latestAlignmentDurationNanos);
+				notifyCheckpoint(receivedBarrier, latestAlignmentDurationNanos);
 			}
-			return false;
+			notifyCheckpointCompletedOrCanceled(barrierId);
```

Comment:
There are five paths for calling `notifyCheckpointCompletedOrCanceled` in this method, then it is a bit hard to trace every path and analysis whether it needs this call. In contrast, we can analysis this issue from a reverse way to make it easy. E.g. only two paths (begin new alignment and under current alignment) do not need the notify call, so we can make a boolean tag only for these two paths, and handle the unified notification at the end of this method.

In detail, we can define a boolean `shouldNotify = true` at the beginning of this method and also set the `notifyCheckpointId = currentCheckpointId`. Then tag this boolean as false in below two paths, and finally call `notifyCheckpointCompletedOrCanceled` at the end of this method if `shouldNotify = true`.

Reply:
For the case of `totalNumberOfInputChannels == 1`, it should be handled separately for fast path. Then I suggest moving this path as a separate method to make #processBarrier seem shorter, and also it can be reused by below `processCancellationBarrier` if providing a function argument.

Reply:
I have changed the implementation